test/randomized/test_transpiler_equivalence.py

```diff
@@ -0,0 +1,225 @@
+
+    @rule(gate=st.sampled_from(oneQ_twoP_gates),
+          qarg=qubits,
+          params=st.lists(
+              st.floats(allow_nan=False, allow_infinity=False),
+              min_size=2, max_size=2))
+    def add_1q2p_gate(self, gate, qarg, params):
+        """Append a random 1q gate with 2 random float parameters."""
+        self.qc.append(gate(*params), [qarg])
+
+    @rule(gate=st.sampled_from(oneQ_threeP_gates),
+          qarg=qubits,
+          params=st.lists(
+              st.floats(allow_nan=False, allow_infinity=False),
+              min_size=3, max_size=3))
+    def add_1q3p_gate(self, gate, qarg, params):
+        """Append a random 1q gate with 3 random float parameters."""
+        self.qc.append(gate(*params), [qarg])
+
+    @rule(gate=st.sampled_from(twoQ_oneP_gates),
+          qargs=st.lists(qubits, max_size=2, min_size=2, unique=True),
+          param=st.floats(allow_nan=False, allow_infinity=False))
+    def add_2q1p_gate(self, gate, qargs, param):
+        """Append a random 2q gate with 1 random float parameter."""
+        self.qc.append(gate(param), qargs)
+
+    @rule(gate=st.sampled_from(twoQ_threeP_gates),
+          qargs=st.lists(qubits, max_size=2, min_size=2, unique=True),
+          params=st.lists(
+              st.floats(allow_nan=False, allow_infinity=False),
+              min_size=3, max_size=3))
+    def add_2q3p_gate(self, gate, qargs, params):
+        """Append a random 2q gate with 3 random float parameters."""
+        self.qc.append(gate(*params), qargs)
+
+    @rule(gate=st.sampled_from(oneQ_oneC_gates),
+          qarg=qubits,
+          carg=clbits)
+    def add_1q1c_gate(self, gate, qarg, carg):
+        """Append a random 1q, 1c gate."""
+        self.qc.append(gate(), [qarg], [carg])
+
+    @rule(gate=st.sampled_from(variadic_gates),
+          qargs=st.lists(qubits, min_size=1, unique=True))
+    def add_variQ_gate(self, gate, qargs):
+        """Append a gate with a variable number of qargs."""
+        self.qc.append(gate(len(qargs)), qargs)
+
+    @precondition(lambda self: len(self.qc.data) > 0)
+    @rule(carg=clbits,
+          data=st.data())
+    def add_c_if_last_gate(self, carg, data):
+        """Modify the last gate to be conditional on a classical register."""
+        creg = carg.register
+        val = data.draw(st.integers(min_value=0, max_value=2**len(creg)-1))
+
+        self.qc.data[-1][0].c_if(creg, val)
+
+    # Properties to check
+
+    @invariant()
+    def qasm(self):
+        """After each circuit operation, it should be possible to build QASM."""
+        self.qc.qasm()
+
+    @precondition(lambda self: any(isinstance(d[0], Measure) for d in self.qc.data))
+    @rule(
+        backend=st.one_of(
+            st.none(),
+            st.sampled_from(backends)),
+        opt_level=st.one_of(
+            st.none(),
+            st.integers(min_value=0, max_value=3)))
+    def equivalent_transpile(self, backend, opt_level):
```

Comment:
`equivalent_transpile` will be invoked at random intervals, like the other methods on `QCircuitMachine`.  I believe invariants can have preconditions, (`equivalent_transpile` was initially also an `@invariant`, I removed it because it slowed the process down substantially. When running in CI, or with operator or measurement-sampled simulations, this wouldn't be an issue though.)